### DISCUSSION TITLE
Custom Nginx config for MatterMost setup

### DISCUSSION THREAD

**Discussion by roland-d** [NginxProxyManager/nginx-proxy-manager#3056] [Q&A] [answered]
Hello,

Running NPM has been a treat, generally I can get it all to work as I want to so it is fun tinkering with stuff. There is one thing I am stuck with and that is configuring Mattermost with NPM.

I have a VPS where all sites run in Docker containers and they are all managed via NPM for their hostname/SSL. So also Mattermost and it works for the most part. There is just an issue with the websocket it seems because the app tells me that notifications cannot be received from the server.

The [documentation from Mattermost](https://docs.mattermost.com/install/installing-ubuntu-2004-LTS.html#id8) has code for how Nginx should be configured. Now with NPM it works a little different as far as I can tell. 

The documentation has this config block for websockets:
```
location ~ /api/v[0-9]+/(users/)?websocket$ {
       proxy_set_header Upgrade $http_upgrade;
       proxy_set_header Connection "upgrade";
       client_max_body_size 50M;
       proxy_set_header Host $http_host;
       proxy_set_header X-Real-IP $remote_addr;
       proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
       proxy_set_header X-Forwarded-Proto $scheme;
       proxy_set_header X-Frame-Options SAMEORIGIN;
       proxy_buffers 256 16k;
       proxy_buffer_size 16k;
       client_body_timeout 60;
       send_timeout 300;
       lingering_timeout 5;
       proxy_connect_timeout 90;
       proxy_send_timeout 300;
       proxy_read_timeout 90s;
       proxy_http_version 1.1;
       proxy_pass http://backend;
   }
```
Now when I put this in the Advanced tab the site won't even come up. Not a real surprise but this is where I am stuck. How should I configure this so it works with NPM?

I do have the Websockets turned on for this host but I guess that alone is not enough.

Any help will be greatly appreciated.

**Comment by GerritKuilder** [comment]
Hi,

this is what I have in my advanced tab:
```
gzip on;
gzip_disable "msie6";
gzip_vary on;
gzip_proxied any;
gzip_comp_level 6;
gzip_buffers 16 8k;
gzip_http_version 1.1;
gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;

proxy_request_buffering off;
proxy_buffering off;
proxy_redirect off;
tcp_nodelay on;
```
And I have one custom location: ^~ /api/v1/streaming 

location: ip of mastodon server ip: 4000

This was all I needed,

Hope this Helps

Gerrit

**Comment by BobWs** [comment]
> Hello,
> 
> 
> 
> Running NPM has been a treat, generally I can get it all to work as I want to so it is fun tinkering with stuff. There is one thing I am stuck with and that is configuring Mattermost with NPM.
> 
> 
> 
> I have a VPS where all sites run in Docker containers and they are all managed via NPM for their hostname/SSL. So also Mattermost and it works for the most part. There is just an issue with the websocket it seems because the app tells me that notifications cannot be received from the server.
> 
> 
> 
> The [documentation from Mattermost](https://docs.mattermost.com/install/installing-ubuntu-2004-LTS.html#id8) has code for how Nginx should be configured. Now with NPM it works a little different as far as I can tell. 
> 
> 
> 
> The documentation has this config block for websockets:
> 
> ```
> 
> location ~ /api/v[0-9]+/(users/)?websocket$ {
> 
>        proxy_set_header Upgrade $http_upgrade;
> 
>        proxy_set_header Connection "upgrade";
> 
>        client_max_body_size 50M;
> 
>        proxy_set_header Host $http_host;
> 
>        proxy_set_header X-Real-IP $remote_addr;
> 
>        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
> 
>        proxy_set_header X-Forwarded-Proto $scheme;
> 
>        proxy_set_header X-Frame-Options SAMEORIGIN;
> 
>        proxy_buffers 256 16k;
> 
>        proxy_buffer_size 16k;
> 
>        client_body_timeout 60;
> 
>        send_timeout 300;
> 
>        lingering_timeout 5;
> 
>        proxy_connect_timeout 90;
> 
>        proxy_send_timeout 300;
> 
>        proxy_read_timeout 90s;
> 
>        proxy_http_version 1.1;
> 
>        proxy_pass http://backend;
> 
>    }
> 
> ```
> 
> Now when I put this in the Advanced tab the site won't even come up. Not a real surprise but this is where I am stuck. How should I configure this so it works with NPM?
> 
> 
> 
> I do have the Websockets turned on for this host but I guess that alone is not enough.
> 
> 
> 
> Any help will be greatly appreciated.

Did you find a working solution?

**Comment by roland-d** [comment]
@GerritKuilder Thank you for your input. However I think this is not related to the websocket part.

@BobWs No, I have not solved this issue yet and am just dealing with it. Just not enough time to dig deeper into this. It seems there is no easy solution for this at least.

**Reply by BobWs** [reply]
@roland-d Okay, I'm using DOCKER Mattermost with NPM on a Synology and it is working just fine! I don't have anything extra added to the NPM configuration...just make sure if you are on the latest Mattermost image that you need to add `*` to `"AllowCorsFrom": "*",` otherwise you will get an error message about the websocket connection.

If you have any questions please let me know, I will try to help if I can.

**Comment by roland-d** [comment]
Hello @BobWs Thank you for your offer to help. Actually Mattermost works fine, it is just that the Mattermost app on Android keeps telling me this:
```
Notifications cannot be received from this server

Due to the configuration for this server, notifications cannot be received in the mobile app. Contact your system admin for more information.
``` 

In practice this means, I need to manually check for new messages. It would be nice to get notified :)

While looking into this, that is when I ended up with the websocket documenation that I mentioned in my opening post.

You are suggesting to add the AllowCorsFrom, so I set this in the config:
![image](https://github.com/NginxProxyManager/nginx-proxy-manager/assets/359377/df476d53-88cb-4b70-a168-17e9f7d5205e)

However the notice remains in the app. Any other suggestions?

**Comment by BobWs** [comment]
Well I was having the same message when I updated my server to version 8.* and the CORS suggestion worked for me. Although I must say I don't use Mattermost on Android, but I guess the problem was the same for IOS and Android.
Have you tried:
- restarting the server
- logout on the Android device and login again
- What is the log saying, maybe you could try to set the log to debug?
Just a few things I can think of what worked for me!

**Comment by roland-d** [comment]
Hello @BobWs 

At the time I was running version 7.x and did not realize there was a newer version. So I updated to 8.0 and the issue remains, I just updated to version 8.1 and as you can guess, the issue still remains.

You mentioned the logs and this made me go through the settings again and I noticed tha tthe `Enable Push Notifications` is turned off. Is that the same for you?

Turning it on, only gives the option to use the push-test server of Mattermost. I have that turned on but see no change for now but let's see if things will change.

**Reply by BobWs** [reply]
Hi @roland-d 
Yes my Push Notification is enabled. See screenshot

![Screen_Shot](https://github.com/NginxProxyManager/nginx-proxy-manager/assets/5485820/add1d82a-1d09-4b55-af8a-ea5140c70345)

**Comment by roland-d** [comment]
@BobWs So I have been giving this some time to see if it needed time to start working but to no avail :) Now I wonder if I need to open up more ports to the Mattermost container. Currently I am only redirecting to port 8065. Do you have any custom locations set up?

**Reply by BobWs** [reply]
Hi, @roland-d no I don't have any other ports open in my router except port 443. The reverse proxy takes care of the routing. There is no need to open any specific port for Mattermost in your router. And I don't have any custom locations setup.
If that's what you mean!

Have you checked your firewall settings?
Have you checked your domain dns settings?
Maybe there is something wrong with on of those settings, I don't know just guessing...

**Comment by BobWs** [comment]
Hi @roland-d 
I just remembered that I had also changed the setting for mobile push notifications before it worked for me too.
Have you looked at these settings yet? See screenshots!

![Screen_Shot](https://github.com/NginxProxyManager/nginx-proxy-manager/assets/5485820/5eff89e1-63ae-451f-9b55-5b5c49589ac6)

![Screen_Shot 1](https://github.com/NginxProxyManager/nginx-proxy-manager/assets/5485820/3d773118-b2ae-4961-b13e-6c243c8386a9)

![Screen_Shot 2](https://github.com/NginxProxyManager/nginx-proxy-manager/assets/5485820/722ce8cf-370e-4af5-ab9c-b5447a6b8a95)

**Comment by roland-d** [comment]
Hello @BobWs Thank you for sticking with this. These settings I had not set yet. Rather, I never changed the default. Now changing it to what you have I am actually getting notified in all possible ways.

The server in the app still has the red triangle saying that `Notifications cannot be reached from this server` however I am receiving notifications. So I guess it is some sort of check that fails but I don´t mind that. Getting the messages is more important.

Once more thank you for all your help.

**Reply by BobWs** [reply, marked as answer]
Hi @roland-d no problem, good to hear that it is working now!

About the warning triangle I had that too, to get rid of that you just need to reinstall the (mobile)app(s) and that will re-register the push-notification server and if I'm right the warning triangle should be gone.

**Comment by roland-d** [comment]
@BobWs You are the hero of the day :) :superhero:  The triangle is gone now as well. Thank you very much for seeing this trough.

**Reply by BobWs** [reply]
Your welcome! Have a good one!
